src/main/java/undo/UndoController.java

```diff
@@ -0,0 +1,84 @@
+package undo;
+
+import backend.Logic;
+import backend.resource.TurboIssue;
+import javafx.application.Platform;
+import javafx.geometry.Insets;
+import javafx.scene.control.Label;
+import javafx.util.Pair;
+import ui.NotificationController;
+import ui.components.Notification;
+import undo.actions.Action;
+import util.DialogMessage;
+
+import java.util.Optional;
+
+public class UndoController {
+
+    @SuppressWarnings("SpellCheckingInspection")
+    private static final String OCTICON_INFO = "\uf059";
+
+    private final Logic logic;
+    private final NotificationController notificationController;
+    private Optional<Pair<TurboIssue, Action<TurboIssue>>> undoBuffer;
+
+    // Holds an undo buffer of size 1.
+    // Buffer stores the undo action.
+    public UndoController(Logic logic, NotificationController notificationController) {
+        this.logic = logic;
+        this.notificationController = notificationController;
+        undoBuffer = Optional.empty();
+    }
+
+    // Gets called when an action is undone
+    public void undoCallback() {
+        if (undoBuffer.isPresent()) {
+            determineLogicMethod(undoBuffer.get().getKey(), undoBuffer.get().getValue());
+            undoBuffer = Optional.empty();
+        }
+    }
+
+    // Adds an action to the undo buffer.
+    @SuppressWarnings("unchecked")
+    public void addAction(TurboIssue issue, Action action) {
```

Comment:
We should use Action<TurboIssue> here, since this method only deals with issues (we can generalise this in future). Then we don't need the cast below and the SuppressWarnings above.